anymail/webhooks/postal.py

```diff
@@ -0,0 +1,175 @@
+    from cryptography.hazmat.primitives.asymmetric import padding
+    from cryptography.exceptions import InvalidSignature
+except ImportError:
+    # This module gets imported by anymail.urls, so don't complain about cryptography missing
+    # unless one of the Postal webhook views is actually used and needs it
+    error = _LazyError(AnymailImproperlyInstalled(missing_package='cryptography', backend='postal'))
+    serialization = error
+    hashes = error
+    default_backend = error
+    padding = error
+    InvalidSignature = object
+
+
+class PostalBaseWebhookView(AnymailBaseWebhookView):
+    """Base view class for Postal webhooks"""
+
+    esp_name = "Postal"
+
+    warn_if_no_basic_auth = False
+
+    # These can be set from kwargs in View.as_view, or pulled from settings in init:
+    webhook_key = None
+
+    def __init__(self, **kwargs):
+        self.webhook_key = get_anymail_setting('webhook_key', esp_name=self.esp_name, kwargs=kwargs, allow_bare=True)
+
+        super().__init__(**kwargs)
+
+    def validate_request(self, request):
+        try:
+            signature = request.META["HTTP_X_POSTAL_SIGNATURE"]
+        except KeyError:
+            raise AnymailWebhookValidationFailure("X-Postal-Signature header missing from webhook")
+
+        public_key = serialization.load_pem_public_key(
+            ('-----BEGIN PUBLIC KEY-----\n' + self.webhook_key + '\n-----END PUBLIC KEY-----').encode(),
+            backend=default_backend()
+        )
+
+        try:
+            public_key.verify(
+                b64decode(signature),
+                request.body,
+                padding.PKCS1v15(),
+                hashes.SHA1()
+            )
+        except InvalidSignature:
+            raise AnymailWebhookValidationFailure(
+                "Postal webhook called with incorrect signature")
+
+
+class PostalTrackingWebhookView(PostalBaseWebhookView):
+    """Handler for Postal message, engagement, and generation event webhooks"""
+
+    signal = tracking
+
+    def parse_events(self, request):
+        esp_event = json.loads(request.body.decode("utf-8"))
+
+        raw_timestamp = esp_event.get("timestamp")
+        timestamp = (
+            datetime.fromtimestamp(int(raw_timestamp), tz=utc)
+            if raw_timestamp
+            else None
+        )
+
+        payload = esp_event.get("payload", {})
+
+        status_types = {
+            "Sent": EventType.DELIVERED,
+            "SoftFail": EventType.DEFERRED,
+            "HardFail": EventType.FAILED,
+            "Held": EventType.QUEUED,
+        }
+
+        def get_event_type(event):
+            if "status" in event:
+                return status_types.get(event["status"], EventType.UNKNOWN)
+
+            if "bounce" in event:
+                return EventType.BOUNCED
+
+            if "url" in event:
+                return EventType.CLICKED
+
+            return EventType.UNKNOWN
+
+        event_type = get_event_type(payload)
```

Comment:
Rather than defining and calling a local function, could I suggest just using `if ... elif ... elif ... else ... ` here to assign `event_type`?

Reply:
Done!